File: c++/devices/th02.cpp
// TH02 Temperature/Humidity Sensor services

// Copyright (C)2018, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <unistd.h>

#include <exception-libsimpleio.h>
#include <th02.h>

// TH02 register addresses

#define RegStatus	0x00
#define RegData		0x01
#define RegConfig	0x03
#define RegID		0x11

// TH02 command codes

#define cmdInit		0x00  // Turn off heater
#define cmdTemp		0x11  // Request temperature sample
#define cmdHumid	0x01  // Request humidity sample

// TH02 status masks

#define Busy		0x01  // Zero when conversion complete

// Device_Class constructor

TH02::Device_Class::Device_Class(Interfaces::I2C::Bus bus)
{
  // Validate parameters

  if (bus == nullptr) THROW_MSG("The bus parameter is NULL");

  // Save some parameters

  this->bus = bus;
  this->addr = 0x40;

  // Turn the heater off

  this->WriteRegister(RegConfig, cmdInit);
}

// Read from an TH02 device register

uint8_t TH02::Device_Class::ReadRegister(uint8_t reg)
{
  uint8_t cmd[1];
  uint8_t resp[1];

  // Validate parameters

  if (reg > RegID)
    THROW_MSG("The reg parameter is out of range");

  if ((reg > RegConfig) && (reg < RegID))
    THROW_MSG("The reg parameter is out of range");

  if ((reg > RegConfig) && (reg < RegID))
    THROW_MSG("The reg parameter is out of range");

  // Build the command

  cmd[0] = reg;

  // Issue the command

  this->bus->Transaction(this->addr, cmd, sizeof(cmd), resp, sizeof(resp));

  // Return the result

  return resp[0];
}

// Write to an TH02 device register

void TH02::Device_Class::WriteRegister(uint8_t reg, uint8_t data)
{
  uint8_t cmd[2];

  // Validate parameters

  if (reg > RegID)
    THROW_MSG("The reg parameter is out of range");

  if ((reg > RegConfig) && (reg < RegID))
    THROW_MSG("The reg parameter is out of range");

  if ((reg > RegConfig) && (reg < RegID))
    THROW_MSG("The reg parameter is out of range");

  // Build the command

  cmd[0] = reg;
  cmd[1] = data;

  // Issue the command

  this->bus->Transaction(this->addr, cmd, sizeof(cmd), nullptr, 0);
}

uint16_t TH02::Device_Class::ReadData(uint8_t which)
{
  uint8_t cmd[1];
  uint8_t resp[2];

  // Start conversion

  this->WriteRegister(RegConfig, which);

  // Wait for completion

  while (this->ReadRegister(RegStatus) & Busy);

  // Fetch result

  cmd[0] = RegData;
  this->bus->Transaction(this->addr, cmd, sizeof(cmd), resp, sizeof(resp));

  // Return result

  return resp[0]*256 + resp[1];
}

// Aquire a temperature sample

double TH02::Device_Class::temperature(void)
{
  return (ReadData(cmdTemp) >> 2)/32.0 - 50.0;
}

// Humidity linearization Coefficients

#define A0 (-4.7844)
#define A1 0.4008
#define A2 (-0.00393)

#define Q0 0.1973
#define Q1 0.00237

// Acquire a humidity sample

double TH02::Device_Class::humidity(void)
{
  // Get humidity sample

  double RHvalue = (ReadData(cmdHumid) >> 4)/16.0 - 24.0;

  // Perform linearization

  double RHlinear = RHvalue - (RHvalue*RHvalue*A2 + RHvalue*A1 + A0);

  // Perform temperature compensation

  return RHlinear + (this->temperature() - 30.0)*(RHlinear*Q1 + Q0);
}

// Fetch the device ID

uint8_t TH02::Device_Class::deviceID(void)
{
  return this->ReadRegister(RegID);
}

File: csharp/devices/th02.cs
// Copyright (C)2018, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

namespace IO.Devices.TH02
{
    /// <summary>
    /// Encapsulates the TH02 temperature and humidity sensor.
    /// </summary>
    public class Device : IO.Interfaces.Temperature.Sensor,
        IO.Interfaces.Humidity.Sensor
    {
        private readonly IO.Interfaces.I2C.Device dev;
        private byte[] cmd = { 0, 0, 0 };
        private byte[] resp = { 0, 0 };

        // TH02 register addresses

        private const byte regStatus = 0x00;
        private const byte regDataH = 0x01;
        private const byte regDataL = 0x02;
        private const byte regConfig = 0x03;
        private const byte regID = 0x11;

        // TH02 commands

        private const byte cmdInit = 0x00;  // Turn heater off
        private const byte cmdTemp = 0x11;
        private const byte cmdHumid = 0x01;

        // TH02 status masks

        private const byte mskBusy = 0x01;  // Nonzero during conversion

        // TH02 humidity correction coefficients (from the TH02 datasheet)

        private const double A0 = -4.7844;
        private const double A1 = 0.4008;
        private const double A2 = -0.00393;

        private const double Q0 = 0.1973;
        private const double Q1 = 0.00237;

        /// <summary>
        /// Constructor for an TH02 temperature and humidity sensor object.
        /// </summary>
        /// <param name="bus">I<sup>2</sup>C bus controller.</param>
        public Device(IO.Interfaces.I2C.Bus bus)
        {
            dev = new IO.Interfaces.I2C.Device(bus, 0x40);
            Write(regConfig, cmdInit);
        }

        // Read from an TH02 device register.

        private byte Read(byte reg)
        {
            // Validate parameters

            if (reg > regID)
                throw new System.Exception("Invalid register address");

            if ((reg > regConfig) && (reg < regID))
                throw new System.Exception("Invalid register address");

            // Issue I2C transaction

            cmd[0] = reg;
            dev.Transaction(cmd, 1, resp, 1);
            return resp[0];
        }

        // Write to an TH02 device register.

        private void Write(byte reg, byte data)
        {
            // Validate parameters

            if (reg > regID)
                throw new System.Exception("Invalid register address");

            if ((reg > regConfig) && (reg < regID))
                throw new System.Exception("Invalid register address");

            // Issue I2C transaction

            cmd[0] = reg;
            cmd[1] = data;
            dev.Write(cmd, 2);
        }

        // Get raw sample data

        private ushort Sample(byte what)
        {
          // Start conversion

          Write(regConfig, what);

          // Wait for completion

          while ((Read(regStatus) & mskBusy) != 0);

          // Fetch result

          cmd[0] = regDataH;
          dev.Transaction(cmd, 1, resp, 2);

          // Return result

          return (ushort)(resp[0]*256 + resp[1]);
        }

        /// <summary>
        /// Read-only property returning the temperature in degrees Celsius.
        /// </summary>
        public double Celsius
        {
            get
            {
                return (Sample(cmdTemp) >> 2) / 32.0 - 50.0;
            }
        }

        /// <summary>
        /// Read-only property returning the temperature in Kelvins.
        /// </summary>
        public double Kelvins
        {
            get
            {
                return IO.Interfaces.Temperature.Conversions.CelsiusToKelvins(Celsius);
            }
        }

        /// <summary>
        /// Read-only property returning the temperature in degrees Fahrenheit.
        /// </summary>
        public double Fahrenheit
        {
            get
            {
                return IO.Interfaces.Temperature.Conversions.CelsiusToFahrenheit(Celsius);
            }
        }

        /// <summary>
        /// Read-only property returning the percentage relative humidity.
        /// </summary>
        public double Humidity
        {
            get
            {
                // Get humidity sample

                double RHvalue = (Sample(cmdHumid) >> 4)/16.0 - 24.0;

                // Perform linearization

                double RHlinear = RHvalue -
                  (RHvalue*RHvalue*A2 + RHvalue*A1 + A0);

                // Perform temperature compensation

                return RHlinear + (Celsius - 30.0)*(RHlinear*Q1 + Q0);
            }
        }

        /// <summary>
        /// Read-only property returning the device ID.
        /// </summary>
        public byte DeviceID
        {
            get
            {
                return Read(regID);
            }
        }
    }
}
